source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,195 @@
+---
+title: Multipath Alerts
+category: feature
+authors: Fred Rolland
+feature_name: Multipath Alerts
+feature_modules: engine,vdsm
+feature_status: Design
+---
+
+# Multipath Alerts
+
+
+## Overview
+
+This feature adds the ability to display alerts to the user in
+the engine UI when some multipath device have faulty paths, when
+some multipath devices are non-operational because all paths became faulty,
+and when all multipath devices on a host recover.
+
+The purpose of this feature is to help users troubleshoot storage issues
+before they become critical when a multipath device is lost, and to
+make it easier to troubleshoot cases when some LUNs are not available
+because all paths became faulty.
+
+
+## Owner
+
+- Name: [Fred Rolland](https://github.com/rollandf)
+- Email: <frolland@redhat.com>
+
+
+## Detailed Description
+
+
+### How it works
+
+Vdsm will register to udev multipath events during storage initialization.
+Vdsm will also build a map of all the faulty paths. 
+For each event, Vdsm will update the list, according to the event.
+The details of the faulty paths will be part of the statistics provided by
+'Host.getStats' verb.
+
+Engine polls the hosts for the statistics every 15 seconds.
+Engine creates an audit log event according to the information provided
+in the statistics. Only if there are changes in the faulty paths report,
+an event will be created.
+
+
+### New vdsm API
+
+New property in 'HostStats': 'MultipathAlertsMap' which is a mapping of faulty paths
+indexed by the device mapper GUID.
+
+
+#### Multipath Alerts Map
+
+The key of the map is the device mapper GUID.
+The value is a MultipathAlert that contains:
+- List of faulty paths.
+- Number of valid paths
+
+### Vdsm multipath event listener
+
+Introduce a new Vdsm internal service for listening to multipath events:
+
+- Code interested in receiving multipath events will subscribe a callback to receive the event
+- When an event is received, all subscribers will receive the events (in another thread).
+- If nobody is registered for events, the events are dropped
+- The multipath event monitor will be started during storage initialization
+
+
+A Udev event monitor is introduced, listening to device-mapper udev events.
+The relevant events for this feature are PATH_FAILED and PATH_REINSTATED:
+
+Variable Name: DM_ACTION
+Uevent Action(s): KOBJ_CHANGE
+Type: string
+Description:
+Value: Device-mapper specific action that caused the uevent action.
+	PATH_FAILED - A path has failed.
+	PATH_REINSTATED - A path has been reinstated.
```

Comment:
I think we can replace the above paragraph (line 72-81) with:

The service will be implemented using pyudev, listening to these events:

- action=change,DM_ACTION=PATH_FAILED
- action=change,DM_ACTION=PATH_REINSTATED
- action=remove (for multiapth devices)

We don't know yet how to detect multiapth devices events efficiently for detecting removed devices.